ENH: Rewrite & cover with tests libcpp.string declarations

zyv opened the pull request:
I did my best to fully sync `libcpp.string` with the standard and cover all overloads with meaningful tests. The only missing bits are the search functions. The highlight of this version is the introduction of iterators and corresponding overloads (note some of the hack comments I left). If anything, this is an excellent stress test for Cython overload resolution engine :-)

robertwb commented:
Is there any rational behind the re-ordering of the declarations (and overload numbering) short of making it complete? All these tests will be really nice to have.

zyv commented:
Hi Robert,

Thanks for having a look! I appreciate your concern that due to the re-ordering the changes are hard to keep track of and I could have potentially introduced some bugs, which would be now hard to see, i.e. mistakenly removed some valid declaration that was already relied upon by the users, etc.

However, let me explain the rationale behind it and why I didn't see any other way forward. The problem with the old ordering is that it is arbitrary, and the overloads are not numbered as they are in the standard, so one is left to guess, whether a specific overload is really what one thinks it is, or maybe it was introduced to account for default parameters handling, and so on. Since the ordering is arbitrary, for every function defined in the standard you have to search whether it's there or not, as opposed to just opening files side by side and comparing the contents.

I tried to complete the class without reordering, but I almost went mad, and in the end decided to reorder everything and cover all the methods with tests to be sure. So, actually, I would like to see the whole standard library re-ordered, because otherwise it's virtually not possible to keep track of standard, and now that it's moving again, I anticipate the need for it in the future.

To address your concerns, I would suggest the following. I will sort both the old definitions and the new definitions and make a whitespace ignoring diff, so that you can easily see what has changed, and post it to this review.

Would this work for you?

robertwb commented:
Yes, that would be helpful. Also useful would be a (long-lived) link to the
standard you're looking at (e.g. in the sources itself).

On Sun, Feb 21, 2016 at 2:08 AM, Yury V. Zaytsev <notifications@github.com>
wrote:

> Hi Robert,
>
> Thanks for having a look! I appreciate your concern that due to the
> re-ordering the changes are hard to keep track of and I could have
> potentially introduced some bugs, which would be now hard to see, i.e.
> mistakenly removed some valid declaration that was already relied upon by
> the users, etc.
>
> However, let me explain the rationale behind it and why I didn't see any
> other way forward. The problem with the old ordering is that it is
> arbitrary, and the overloads are not numbered as they are in the standard,
> so one is left to guess, whether a specific overload is really what one
> thinks it is, or maybe it was introduced to account for default parameters
> handling, and so on. Since the ordering is arbitrary, for every function
> defined in the standard you have to search whether it's there or not, as
> opposed to just opening files side by side and comparing the contents.
>
> I tried to complete the class without reordering, but I almost went mad,
> and in the end decided to reorder everything and cover all the methods with
> tests to be sure. So, actually, I would like to see the whole standard
> library re-ordered, because otherwise it's virtually not possible to keep
> track of standard, and now that it's moving again, I anticipate the need
> for it in the future.
>
> To address your concerns, I would suggest the following. I will sort both
> the old definitions and the new definitions and make a whitespace ignoring
> diff, so that you can easily see what has changed, and post it to this
> review.
>
> Would this work for you?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cython/cython/pull/488#issuecomment-186789448>.
>

scoder commented:
Any update on this?

zyv commented:
Hi Stefan, thank you for pinging me! I haven't forgotten about it, but, unfortunately, I wasn't able to find another opportunity to work on this & #489 so far, because (sadly) the management decided that they can't grant me the time to wrap it up. I'm still trying to find some time to finish it, and hopefully will get a bit more luck in October... I know this must be very annoying for you, but it's the way it is :-/

scoder commented:
It's your time that you're investing, so thanks for doing that in the first place.

scoder commented:
As expected, the restructuring here has lead to various conflicts with the current master, and given that the diff is not informative, it's unclear what needs to be done to resolve it. Help wanted to get this resolved.

andrewasche commented:
I'm willing to help, but don't know what needs to be done.

scoder commented:
I think the original reordering of the declarations is ok, but the PR now conflicts with master. We need two things:
- validate that we didn't loose any of the declarations in the reordering done in this PR
- integrate the changes from master that created the conflicts back into this PR so that we end up with a complete set of ordered declarations.
- optionally: if there were additions in master, see if we can also cover those with tests.

Any help would be appreciated.

zyv commented:
I've tried to look into rebasing this PR, but I think that's only possible by reconstructing the changes in the declarations. However, I found GitHub's conflict resolution UI for tests to be surprisingly useful.

The biggest problem with the tests was that some tests have been added in the meantime (it's only been a decade or so). Some of them duplicate the ones I added and need to be looked at.

Otherwise, I've replaced many of the `char *` with `string`. I'm not sure why the old tests always used `char *` and converted it to a C++ string inside the tests. Probably it wasn't possible due to missing/non-functional overloads, or maybe no one really thought of that.

In any case, the test file is now at least technically "rebased", but of course the tests might fail if some overloads are still missing. I'll try to take a look at it.

zyv commented:
I've cleaned up the tests:

* Conflicting code:
  * `test_npos`, `test_pop_back` - switched to `string`
  * `test_assign`, `test_erase`, `test_replace`, `test_swap` - my tests were more extensive
* Switched all I could to `string` instead of `char *`
* Removed `c_str()` where possible
* Removed asserts

Apparently the tests are running now, so most of the overloads that were missing were actually accepted - and what wasn't accepted, but was tested by me, I just added.

I could see if I can reapply my work, but as I said, I think anything other than re-odering to the standard is too crazy. I could redo this work, but I'm afraid it will be stalled for the next decade pending review, and by then there will be thousands of conflicts again.

@da-woods @scoder, if I go through the code again, will you take the time to look at it? Any other ideas?

P.S. This is the permanent link to the C++14 standard (draft):

https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3797.pdf

da-woods commented on Cython/Includes/libcpp/string.pxd:
```diff
@@ -178,6 +181,7 @@ cdef extern from "<string>" namespace "std" nogil:
         string& append(const char* s) except +
         string& append(const char* s, size_t n) except +
         string& append(size_t n, char c) except +
+        string_ref append[input_iterator](input_iterator, input_iterator) except +  # (6)
```
What does the `#6` refer to here?

zyv replied:
This refers to the number of the overload in the C++ standard. You will notice that in my rewrite (which I have abandoned for the purposes of the rebase until a consensus is reached, as explained in the comment), I have consistently numbered all overloads according to the standard to ensure that they are implemented correctly and none are missing.

<img width="933" alt="Screenshot 2025-01-21 at 19 10 49" src="https://github.com/user-attachments/assets/17c236e1-3089-4ec9-a931-1919398ab554" />

da-woods reviewed (commented):
Left a few comments. But it mostly looks reasonable.

I should be able to review it if you update it I think.

My inclination would be to try to break into steps that need minimal thought to review (and so can be merged fairly quickly). i.e.
1. merge this (+ some small fixes) just updating the tests
2. new PR re-arranging to the order of the standard but not adding anything
3. new PR with additions.

or something like that. Although possibly splitting steps 2 and 3 is too much effort.

zyv commented on tests/run/cpp_stl_string.pyx:
```diff
@@ -54,234 +55,333 @@ def test_conversion(py_obj):
     cdef string s = py_obj
     return s
 
-def test_indexing(char *py_str):
```
It has been renamed to `test_access`. Sorry, the diff looks stupid. Here is the code for reference:

```python
def test_indexing(char *py_str):
    """
    >>> test_indexing(b_asdf)
    ('s', 's')
    """
    cdef string s
    s = string(py_str)
    return chr(s[1]), chr(s.at(1))
```

```python
def test_access(string s, i):
    """
    >>> test_access(b_asdf, 1) == (ord(b_asdf[1:2]), ) * 2
    True
    >>> test_access(b_asdf, 5)  # doctest: +ELLIPSIS
    Traceback (most recent call last):
        ...
    IndexError: ...
    """
    return s.at(i), s[I]
```

da-woods replied:
That's fine thanks.

zyv commented on Cython/Includes/libcpp/string.pxd:
```diff
@@ -125,7 +128,7 @@ cdef extern from "<string>" namespace "std" nogil:
         string(const char* s, size_t n) except +
         string(size_t n, char c) except +
-        string(iterator first, iterator last) except +
+        string_ref string[input_iterator](input_iterator, input_iterator) except +  # (6)
```
Oh yes, unfortunately templated constructors don't work without type. I've replaced it with `void` if that's any better. Or is `string&` actually clearer? Should I add some kind of comment?

da-woods replied:
There was a comment `        # return type is a hack to allow the code to compile` in one of the diffs along the way. I think it's a useful comment. `string_ref` is probably clearer, but it doesn't matter too much.

zyv replied:
TODO: add comment or remove templating. See #489.

zyv replied:
Done.

zyv commented:
I have now removed the reordering and as far as I'm concerned this is ready for merge. Please let me know if there is anything else I can/should do.

Also, please let me know if there is interest in a follow-up to the re-ordering. I can see if I can (re)do it...

da-woods commented:
Thanks
